misc/agent-introspection-validator/agent-introspection-validator.go

```diff
@@ -0,0 +1,267 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/cihub/seelog"
+)
+
+const (
+	agentIntrospectionEndpoint = "http://localhost:51678/v1/"
+	maxRetries                 = 4
+	durationBetweenRetries     = time.Second
+	containerName              = "agent-introspection-validator"
+	tasksMetadataRespType      = "tasks metadata"
+	taskMetadataRespType       = "task metadata"
+)
+
+// TaskResponse is the schema for the task response JSON object
+type TaskResponse struct {
+	Arn           string              `json:"Arn"`
+	DesiredStatus string              `json:"DesiredStatus,omitempty"`
+	KnownStatus   string              `json:"KnownStatus"`
+	Family        string              `json:"Family"`
+	Version       string              `json:"Version"`
+	Containers    []ContainerResponse `json:"Containers"`
+}
+
+// TasksResponse is the schema for the tasks response JSON object
+type TasksResponse struct {
+	Tasks []*TaskResponse `json:"Tasks"`
+}
+
+// ContainerResponse is the schema for the container response JSON object
+type ContainerResponse struct {
+	DockerID   string         `json:"DockerId"`
+	DockerName string         `json:"DockerName"`
+	Name       string         `json:"Name"`
+	Ports      []PortResponse `json:"Ports,omitempty"`
+	Networks   Network        `json:"Networks,omitempty"`
+}
+
+// PortResponse defines the schema for portmapping response JSON
+// object.
+type PortResponse struct {
+	ContainerPort uint16 `json:"ContainerPort,omitempty"`
+	Protocol      string `json:"Protocol,omitempty"`
+	HostPort      uint16 `json:"HostPort,omitempty"`
+}
+
+// Network is a struct that keeps track of metadata of a network interface
+type Network struct {
+	NetworkMode   string   `json:"NetworkMode,omitempty"`
+	IPv4Addresses []string `json:"IPv4Addresses,omitempty"`
+	IPv6Addresses []string `json:"IPv6Addresses,omitempty"`
+}
+
+func getTasksMetadata(client *http.Client, path string) (*TasksResponse, error) {
+	body, err := metadataResponse(client, path, tasksMetadataRespType)
+	if err != nil {
+		return nil, err
+	}
+
+	seelog.Infof("Received tasks metadata: %s \n", string(body))
+
+	var tasksMetadata TasksResponse
+	err = json.Unmarshal(body, &tasksMetadata)
```

Comment:
Unmarshal could also work with case insensitive, this won't capture the bug, example: https://play.golang.org/p/iH8nsK-UJ_L